mautic/mautic
# Replace fetching updates from updates.mautic.org to use Github releases

**Please be sure you are submitting this against the _staging_ branch.**

[//]: # ( Please answer the following questions: )

| Q  | A
| --- | ---
| Bug fix? | 
| New feature? | Y
| Automated tests included? | Y
| Related user documentation PR URL | 
| Related developer documentation PR URL | 
| Issues addressed (#s or URLs) | 
| BC breaks? | 
| Deprecations? | 

[//]: # ( Note that all new features should have a related user and/or developer documentation PR in their respective repositories. )


[//]: # ( Required: )
#### Description:

Similar to https://github.com/mautic/mautic/pull/8226, this PR is removing another dependency on https://updates.mautic.org by pulling releases directly from Github. It uses Github's public releases API endpoint to get a list of releases. It then checks for a `app/release_metadata.json` file that contains min/max supported PHP version, stability, announcement URL (will use Github's release if empty), and minimum supported Mautic version. It then determines the applicable latest supported release. 

(Note - download statistics can be collected using Github's release API for the assets)

### 
[//]: # ( As applicable: )

#### Steps to test this PR:
1. Install this PR as a new Mautic installation
2. add `    'system_update_url'         => 'https://api.github.com/repos/alanhartless/mautic/releases',` to `app/config/local.php` and clear the cache. JUST KIDDING, you don't have the clear the cache any more for most local.php changes. 🎉
3. Login and note that bottom right has the version as 3.0.0-beta
4. There should be a notification for the 3.0.0 update being available
5. Go through the in-app update
6. Mautic should now show 3.0.0 as the installed version
7.  Repeat with a new installation and this time use the update command `php bin/console mautic:update:find` then `php bin/console mautic:update:apply` and then finally `php bin/console mautic:update:apply --finish`.

#### List backwards compatibility breaks:
1. `\AppKernel::MAJOR_VERSION`, `\AppKernel::MINOR_VERSION`, `\AppKernel::PATCH_VERSION`, and `\AppKernel::EXTRA_VERSION` are no longer defined. Use `\Mautic\CoreBundle\Release\ThisRelease::getMetadata()->getMajorVersion()`, etc instead. 
2. `\Mautic\CoreBundle\Helper\UpdateHelper::getServerOs` is now private
